### PR TITLE
SNOW-1119855: Push base unimplemented methods to query compiler when possible (1/?)

### PR DESCRIPTION
**Pull request by sfc-gh-joshi**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1119855 (first of several PRs)

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

This PR prepares for the imminent removal of base.py from the Snowpark pandas codebase by pushing as many unimplemented methods down to the query compiler as possible. Some unimplemented methods must still be overridden because Modin uses other implemented query compiler/frontend methods in their implementation; these methods have been moved to `src/snowflake/snowpark/modin/plugin/extensions/base_overrides.py`.

### The following methods are affected:

The below BasePandasDataset methods either have existing corresponding query compiler stubs that can be used directly, or default to pandas *at the query compiler layer*, which Snowpark pandas specifies to raise a NotImplementedError.
<details>
<summary><b>Pushed to the query compiler</b></summary>
<ul>
<li>align</li>
<li>at_time (via between_time)</li>
<li>between_time</li>
<li>clip</li>
<li>combine (via binary_op)</li>
<li>combine_first (via binary_op)</li>
<li>explode</li>
<li>infer_objects</li>
<li>kurt</li>
<li>kurtosis (via kurt)</li>
<li>mode</li>
<li>sem</li>
<li>tz_convert (via tz_convert)</li>
<li>tz_localize (via tz_localize)</li>
<li>__sizeof__</li>
</ul>
</details>

The below methods do not have corresponding query compiler methods, and either (1) would work in Snowpark pandas but haven't been tested/need too many queries, (2) use frontend/other query compiler methods that we cannot override, or (3) manipulates index/columns in ways that we may not support.
<details>
<summary><b>Moved to base_overrides.py</b></summary>
<li>asof: modin uses dropna + reindex</li>
<li>bool: method is deprecated in pandas but pretty trivial</li>
<li>droplevel: modin does manipulations on column/index objects</li>
<li>ewm: modin defaults at frontend, would materialize data</li>
<li>filter: modin uses loc + manipulations on column/index</li>
<li>pipe: just kind of weird, probably would materialize data</li>
<li>pop: modin uses loc + del</li>
<li>reorder_levels: modin uses properties on index object</li>
<li>set_flags: modin defaults at frontend, would materialize data</li>
<li>swapaxes: modin uses transpose + index/column manipulations</li>
<li>swaplevel: modin uses native index swaplevel</li>
<li>transform: modin uses QC agg method</li>
<li>truncate: modin manipulates index objects</li>
<li>__finalize__: modin defaults at forntend, would materialize data</li>
</details>

These methods are also moved to base_overrides.py.
<details>
<summary><b>I/O methods that default on the frontend and would materialize data</b></summary>
<li>to_clipboard</li>
<li>to_excel</li>
<li>to_hdf</li>
<li>to_json</li>
<li>to_latex</li>
<li>to_markdown</li>
<li>to_pickle</li>
<li>to_string</li>
<li>to_sql</li>
<li>to_timestamp</li>
<li>to_xarray</li>
</details>

<details>
<summary><b>Other</b></summary>
<li>__array_wrap__: this was removed in pandas 2.0</li>
<li>reindex_like: at some point, this method was moved from base.py into dataframe.py/series.py upstream; we continue to mark these as unimplemented in our vendored copies of dataframe/series for now</li>
<li>map: this was also split into dataframe.py/series.py; since we partially support the method for series (but not currently DF) we mark this method as unimplemented only in dataframe.py</li>
<li>transform: we partially support the method for DF (but not for series), so we add an override in dataframe_overrides.py with an implementation and an unimplemented stub in series_overrides.py </li>
</details>

### CODE REVIEW

**Comment by sfc-gh-azhan** [comment]
You can move base.py out of the code coverage check given it will be removed soon.

File: src/snowflake/snowpark/modin/pandas/base.py [hunk 1]
```diff
@@ -1722,7 +1740,6 @@ def eq(self, other, axis="columns", level=None):  # noqa: PR01, RT01, D200
         # TODO: SNOW-1119855: Modin upgrade - modin.pandas.base.BasePandasDataset
         return self._binary_op("eq", other, axis=axis, level=level, dtypes=np.bool_)
 
-    @base_not_implemented()
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/modin/pandas/base.py:1725]: what happens now if we remove it?

**Review by sfc-gh-azhan — APPROVED** [review]

**Comment by sfc-gh-joshi** [comment]
> You can move base.py out of the code coverage check given it will be removed soon.

If I modify tox.ini (which seems to be where files get excluded) it'll trigger a review request on the Python team, so I'll just add manual `# pragma: no cover`s for now.

**Review by sfc-gh-rdurrani — APPROVED** [review]
LGTM, great work Jonathan!

**Pull request merged**